java/ql/src/experimental/Security/CWE/CWE-094/JexlInjectionLib.qll

```diff
@@ -0,0 +1,464 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+
+/**
+ * A taint-tracking configuration for unsafe user input
+ * that is used to construct and evaluate a Jexl expression.
+ * It supports both Jexl2 and Jexl3.
+ */
+class JexlInjectionConfig extends TaintTracking::Configuration {
+  JexlInjectionConfig() { this = "JexlInjectionConfig" }
+
+  override predicate isSource(DataFlow::Node source) {
+    source instanceof TaintedSpringRequestBody or
+    source instanceof RemoteFlowSource or
+    source instanceof UserInput or
+    source instanceof EnvInput
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof JexlEvaluationSink }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    creatingTaintedJexlExpression(node1, node2) or
+    creatingTaintedJexlTemplate(node1, node2) or
+    creatingTaintedJexlScript(node1, node2) or
+    creatingTaintedJexlCallable(node1, node2) or
+    returningTaintedDataFromBean(node1, node2)
+  }
+}
+
+/**
+ * A data flow source for parameters that have
+ * a Spring framework annotation indicating remote user input from servlets.
+ */
+class TaintedSpringRequestBody extends DataFlow::Node {
+  TaintedSpringRequestBody() {
+    exists(SpringServletInputAnnotation a | this.asParameter().getAnAnnotation() = a)
+  }
+}
+
+/**
+ * A sink for Expresssion Language injection vulnerabilities via Jexl,
+ * i.e. methods that run evaluation of a Jexl expression.
+ */
+class JexlEvaluationSink extends DataFlow::ExprNode {
+  JexlEvaluationSink() {
+    isJexlExpressionEvaluationCall(asExpr()) or
+    isJexlTemplateEvaluationCall(asExpr()) or
+    isJexlScriptExecuteCall(asExpr()) or
+    isJexlGetSetPropertyCall(asExpr()) or
+    isCallableCall(asExpr())
+  }
+}
+
+/**
+ * Holds if `node1` to `node2` is a dataflow step that creates a Jexl expression.
+ */
+predicate creatingTaintedJexlExpression(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma, Method m | ma.getMethod() = m |
+    (
+      m instanceof JxltEngineCreateExpressionMethod or
+      m instanceof UnifiedJexlParseMethod or
+      m instanceof JexlEngineCreateExpressionMethod
+    ) and
+    ma.getAnArgument().getType() instanceof TypeString and
+    ma.getAnArgument() = node1.asExpr() and
+    node2.asExpr() = ma
+  )
+}
+
+/**
+ * Holds if `node1` to `node2` is a dataflow step that creates a Jexl expression.
+ */
+predicate creatingTaintedJxltEngineExpression(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma, Method m | ma.getMethod() = m |
+    (m instanceof JxltEngineCreateExpressionMethod or m instanceof UnifiedJexlParseMethod) and
```

Comment:
Good catch. At first, the query supported only JEXL 3. Then, I added support for JEXL 2. The functionality in those two versions is almost the same, but the API are slightly different. I didn't want to have separate sinks and taint propagating steps for the two versions because it would result in a lot of dup code. Currently, the sinks and steps cover both JEXL 2 and 3. But I forgot to remove some redundand stuff. Thanks!